### PR TITLE
Feature/application

### PR DESCRIPTION
**Pull request by lebdron**
## What is this pull request?
Concrete class that contains instances of components which are shared in the system
   
## Why do you implement it? Why do we need this pull request?
It will be used in main application during system initialization, used by Torii, and other components
  
## How to use the features provided in the pull request?
Instantiate this class in main application, handle the errors and ensure the correctness of component initialization, and pass it to other components

## Details/Features
Application class which contains following instances:
* Ametsuchi storage
* Peer communication service
* Crypto provider
* Hash provider

### CODE REVIEW

**Comment by motxx** [comment]
Will `Application` be expected to consist of other all modules?

File: irohad/main/application.hpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+#ifndef IROHA_APPLICATION_HPP
+#define IROHA_APPLICATION_HPP
+
+#include <network/network_api.h>
+#include <ametsuchi/ametsuchi.hpp>
+#include <dao/dao.hpp>
+
+namespace iroha {
```
Comment (MizukiSonoko) [inline irohad/main/application.hpp]: Why Is `iroha` namespace used?

File: irohad/main/application.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  /**
+   * Contains instances of global services and shared providers
+   */
+  class Application {
```
Comment (MizukiSonoko) [inline irohad/main/application.hpp]: Why Is Application class?
I think main contains `int main(){};` basically, 
In my opinion, Application class is a little fat...

Comment (muratovv) [inline irohad/main/application.hpp]: Application is bundle with all services. Torii module know about other in general. But parts of torii will consume part of application directly.

File: irohad/main/application.hpp (hunk 1 above)
Comment (muratovv) [inline irohad/main/application.hpp]: What is your proposal for this?

**Comment by muratovv** [comment]
@motxx idea is global modules knows about other modules. Each module by itself separate services by responsibilities  and provide only part of interface.

For example, stateful validation should know about ametsuchi?
`Proposal validate(Proposal, Ametsuchi)`
But this is bad interface. For performing validation you should use only temporary database state (without affecting on blockstore) - interface will be:
`Proposal validate(Proposal, TemporaryWsv)`

We reduce responsibility of validate interface.

For using it in business logic you should invoke like this:
`// peer communication class`
`auto verified_proposal = validate(proposal, ametsuchi.getTemporaryWsv())`

So, what happens? Global module PeerCommunicationService know about ametsuchi, but parts like stateful validation works only with small part of it.

**Comment by motxx** [comment]
@muratovv
class PeerCommunicationService : public TransactionPropagator,
                                    public ConsensusListener {};

**Comment by motxx** [comment]
I'm sorry, I pushed the close button by mistake. Looks nice for me.

File: irohad/main/application.hpp (hunk 2 above)
Comment (motxx) [inline irohad/main/application.hpp]: I think the name of `Application` assumes to contain consensus and other modules.
How is `SharedServices`?

File: irohad/main/application.hpp (hunk 1 above)
Comment (MizukiSonoko) [inline irohad/main/application.hpp]: I think this module is `iroha` obviously. and this module is not shared by another library. only iroha.
so I propose 
`namespace main{}`

File: irohad/main/application.hpp (hunk 2 above)
Comment (MizukiSonoko) [inline irohad/main/application.hpp]: ok i see

**Comment by MizukiSonoko** [comment]
In my opinion, and 
**-- In the future iroha, this is only memo --**
If there is application in hyperledger/iroha, There is `class Application` in libs/application. and it has
```
virtual void initialize();
virtual void commandParser(int argc,char* argv);
virtual void run();
...
```

and There is `class Irohad : private Application{}` in irohad/main/irohad.hpp.

I think Application is separated from irohad/main is better. 
but, Is it complex?
**-- --**
or 
I propose we use name `class Irohad` instead of `class Application`

**Comment by muratovv** [comment]
@MizukiSonoko 
I don't mind for renaming. Pls, commit changes in this branch.

**Comment by MizukiSonoko** [comment]
Ok 
Can I rename like this?
`class Irohad` instead of `class Application`

**Comment by muratovv** [comment]
Yes, u are.

**Comment by muratovv** [comment]
@motxx 
Could you review and merge it?

**Pull request merged**